editor/project_manager.cpp

```diff
@@ -443,6 +443,160 @@ class NewProjectDialog : public ConfirmationDialog {
+
+		error->set_text("");
+		get_ok()->set_disabled(true);
+		DirAccess *d = DirAccess::create(DirAccess::ACCESS_FILESYSTEM);
+		String valid_path;
+		if (d->change_dir(project_path->get_text()) == OK) {
+			valid_path = project_path->get_text();
+		} else if (d->change_dir(project_path->get_text().strip_edges()) == OK) {
+			valid_path = project_path->get_text().strip_edges();
+		}
+
+		if (valid_path == "") {
+			error->set_text(TTR("Invalid project path, the path must exist!"));
+			memdelete(d);
+			return "";
+		}
+
+		if (valid_path != "" && !d->file_exists("project.godot")) {
+
+			error->set_text(TTR("Invalid project path, project.godot must exist."));
+			memdelete(d);
+			return "";
+		}
+
+		memdelete(d);
+		get_ok()->set_disabled(false);
+		return valid_path;
+	}
+
+	void _text_changed(const String &p_text) {
+
+		error->set_text("");
+		if (p_text == "") {
+
+			error->set_text(TTR("Name cannot be empty"));
+			get_ok()->set_disabled(true);
+			return;
+		}
+		get_ok()->set_disabled(false);
+	}
+
+	void ok_pressed() {
+
+		String dir = _test_path();
+		if (dir == "") {
+			error->set_text(TTR("Invalid project path (changed anything?)."));
+			return;
+		}
+
+		ProjectSettings *current = memnew(ProjectSettings);
+		current->add_singleton(ProjectSettings::Singleton("Current"));
+
+		if (current->setup(dir, "")) {
+			error->set_text(TTR("Couldn't get project.godot in project path."));
+		} else {
+			ProjectSettings::CustomMap edited_settings;
+			edited_settings["application/config/name"] = project_name->get_text();
+
+			if (current->save_custom(dir.plus_file("/project.godot"), edited_settings, Vector<String>(), true)) {
+				error->set_text(TTR("Couldn't edit project.godot in project path."));
+			}
+		}
+
+		emit_signal("project_renamed");
+	}
+
+protected:
+	static void _bind_methods() {
+
+		ClassDB::bind_method("_text_changed", &RenameProjectDialog::_text_changed);
+		ADD_SIGNAL(MethodInfo("project_renamed"));
+	}
+
+public:
+
+	void set_project_path(const String &p_path) {
+		project_path->set_text(p_path);
+	}
+
+	void show_dialog() {
+
+		set_title(TTR("Rename Project"));
+		get_ok()->set_text(TTR("Rename"));
+
+		String dir = _test_path();
+		if (dir == "") {
+			error->set_text(TTR("Invalid project path (changed anything?)."));
+			return;
+		}
+		ProjectSettings *current = memnew(ProjectSettings);
+		current->add_singleton(ProjectSettings::Singleton("Current"));
+
+		if (current->setup(dir, "")) {
+			error->set_text(TTR("Couldn't get project.godot in project path."));
+		} else {
+			if (current->has("application/config/name")) {
+				String appname = current->get("application/config/name");
+				project_name->set_text(appname);
+			}
+		}
+
+		popup_centered(Size2(500, 125) * EDSCALE);
+		project_name->grab_focus();
+
+	}
+
+	RenameProjectDialog() {
+
+		VBoxContainer *vb = memnew(VBoxContainer);
+		add_child(vb);
+
+		Label *l = memnew(Label);
+		l->set_text(TTR("Project Path:"));
+		vb->add_child(l);
+
+
```

Comment:
Double newlines here. Clang-format won't like that. :P

Reply:
Ups.